File: src/main/java/me/oddlyoko/odoo/modules/OdooModuleResolver.java
package me.oddlyoko.odoo.modules;

import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.QualifiedName;
import com.jetbrains.python.psi.impl.PyImportResolver;
import com.jetbrains.python.psi.resolve.PyQualifiedNameResolveContext;
import com.jetbrains.python.psi.resolve.PyResolveImportUtil;
import me.oddlyoko.odoo.modules.models.OdooModule;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * Import module resolver
 */
public class OdooModuleResolver implements PyImportResolver {
    @Nullable
    @Override
    public PsiElement resolveImportReference(@NotNull QualifiedName name, @NotNull PyQualifiedNameResolveContext context, boolean withRoots) {
        List<String> components = name.getComponents();
        if (components.size() < 3)
            return null;
        if ("odoo".equals(components.get(0)) && "addons".equals(components.get(1))) {
            String moduleName = components.get(2);
            PsiDirectory directory = context.getContainingDirectory();
            if (directory == null)
                return null;
            OdooModule currentModule = OdooModuleUtil.getModuleFromDirectory(directory);
            if (currentModule == null)
                return null;
            OdooModule target = OdooModuleUtil.getModule(moduleName, context.getProject());
            if (target == null)
                return null;
            return PyResolveImportUtil.resolveModuleAt(name.subQualifiedName(3, name.getComponentCount()), target.getDirectory(), context)
                    .stream()
                    .findFirst().orElse(null);
        }
        return null;
    }
}

File: src/main/kotlin/me/oddlyoko/odoo/modules/OdooModuleResolver.kt
package me.oddlyoko.odoo.modules

import com.intellij.psi.PsiElement
import com.intellij.psi.util.QualifiedName
import com.jetbrains.python.psi.impl.PyImportResolver
import com.jetbrains.python.psi.resolve.PyQualifiedNameResolveContext
import com.jetbrains.python.psi.resolve.ResolveImportUtil
import me.oddlyoko.odoo.modules.models.OdooModule

class OdooModuleResolver: PyImportResolver {
    override fun resolveImportReference(name: QualifiedName, context: PyQualifiedNameResolveContext, withRoots: Boolean): PsiElement? {
        val components = name.components
        if (components.size < 3)
            return null
        if ("odoo" == components[0] && "addons" == components[1]) {
            val targetName = components[2]
            val target: OdooModule = OdooModuleUtil.getModule(targetName, context.project) ?: return null

            // 'import odoo.addons.xxx' or 'from odoo.addons.xxx import yyy'
            if (components.size == 3)
                return target.directory

            // 'import odoo.addons.xxx.yyy.zzz' or 'from odoo.addons.xxx.yyy import zzz'
            return resolveOdooAddonImportReference(name.removeHead(3), target, context)
        }
        return null
    }

    private fun resolveOdooAddonImportReference(name: QualifiedName, module: OdooModule, context: PyQualifiedNameResolveContext): PsiElement? {
        var parent: PsiElement? = module.directory
        for (component in name.components)
            parent = ResolveImportUtil.resolveChildren(parent, component, context.footholdFile, !context.withMembers, !context.withPlainDirectories, context.withoutStubs, context.withoutForeign)
                .firstOrNull()?.element
        return parent
    }
}
